In [68]:
import pandas as pd
from sklearn.metrics import accuracy_score, recall_score, precision_score, f1_score
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LogisticRegression

df = pd.read_csv("C:\\Users\In9.var\Desktop\Projects\data\diabetes.csv")
df

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0              6      148             72             35        0  33.6   
1              1       85             66             29        0  26.6   
2              8      183             64              0        0  23.3   
3              1       89             66             23       94  28.1   
4              0      137             40             35      168  43.1   
..           ...      ...            ...            ...      ...   ...   
763           10      101             76             48      180  32.9   
764            2      122             70             27        0  36.8   
765            5      121             72             23      112  26.2   
766            1      126             60              0        0  30.1   
767            1       93             70             31        0  30.4   

     DiabetesPedigreeFunction  Age  Outcome  
0                       0.627   50        1  
1                       0.351   31        0  
2                       0.672   32        1  
3                       0.167   21        0  
4                       2.288   33        1  
..                        ...  ...      ...  
763                     0.171   63        0  
764                     0.340   27        0  
765                     0.245   30        0  
766                     0.349   47        1  
767                     0.315   23        0  

[768 rows x 9 columns]

In [69]:
X = df.drop(["Outcome"], axis=1)
y = df["Outcome"]

from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=1/3, random_state=42)

stat = []
for i in range(1, 10):
    tree = DecisionTreeClassifier(criterion='gini', max_depth=i, random_state=42)
    cvs = cross_val_score(estimator=tree, X=X, y=y, cv=5, scoring='accuracy')
    stat.append((i, cvs.mean()))
    
print(max(stat, key=lambda x: x[1]))

(5, 0.7487819370172311)


In [70]:
#Кросс валидация. Оценка моделей на обычных данных
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold

# дерево

tree = DecisionTreeClassifier(criterion='gini', max_depth=5, random_state=42)

kf = KFold(n_splits=5)

cvs = cross_val_score(estimator=tree, X=X, y=y, cv=5, scoring='accuracy')
cvs1 = cross_val_score(estimator=tree, X=X, y=y, cv=5, scoring='recall')
cvs2 = cross_val_score(estimator=tree, X=X, y=y, cv=5, scoring='precision')
print('Accuracy:', cvs.mean())
print('Recall:', cvs1.mean())
print('Precision:', cvs2.mean())
print('F1_score:', (2 * cvs1.mean() * cvs2.mean())/(cvs1.mean() + cvs2.mean()))




Accuracy: 0.7487819370172311
Recall: 0.5825995807127883
Precision: 0.6614478114478114
F1_score: 0.6195249795807436


In [71]:
# KNN

X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)

knn = KNeighborsClassifier(7)

knn.fit(X_train,y_train)

y_pred = knn.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))






Accuracy: 0.7109375
Recall: 0.5465116279069767
Precision_score: 0.573170731707317
f1: 0.5595238095238095


In [72]:
# логист. регрессия для норм данных


model = LogisticRegression()
model = model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))





Accuracy: 0.7421875
Recall: 0.5348837209302325
Precision_score: 0.6388888888888888
f1: 0.5822784810126581


In [73]:
# создание станд. и норм. данных
from sklearn.preprocessing import StandardScaler, MinMaxScaler

scaler = StandardScaler()
X_st = scaler.fit_transform(X)
scaler = MinMaxScaler()
X_norm = scaler.fit_transform(X)





In [74]:
# Определение лучшего значения для knn

# норм. данные
X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
scores = []
for i in range (1, 15):
    knn = KNeighborsClassifier(i)
    
    knn.fit(X_train, y_train)
    
    scores.append(knn.score(X_test, y_test))
max_score = max(scores)
scores_ind = [i for i, v in enumerate(scores) if v == max_score]
print(scores_ind)


# станд. данные
X_train, X_test, y_train, y_test = train_test_split(X_st, y, test_size=1/3, random_state=42)
scores = []
for i in range (1, 15):
    knn = KNeighborsClassifier(i)
    
    knn.fit(X_train, y_train)
    
    scores.append(knn.score(X_test, y_test))
max_score = max(scores)
scores_ind = [i for i, v in enumerate(scores) if v == max_score]
print(scores_ind)

# обычные данные
X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)
scores = []
for i in range (1, 15):
    knn = KNeighborsClassifier(i)
    
    knn.fit(X_train, y_train)
    
    scores.append(knn.score(X_test, y_test))
max_score = max(scores)
scores_ind = [i for i, v in enumerate(scores) if v == max_score]
print(scores_ind)





[8]
[6]
[8]


In [75]:
# станд. данные
# Соседи
X_train, X_test, y_train, y_test = train_test_split(X_st, y, test_size=1/3, random_state=42)

knn = KNeighborsClassifier(5)

knn.fit(X_train,y_train)

y_pred = knn.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))




Accuracy: 0.6875
Recall: 0.5
Precision_score: 0.5375
f1: 0.5180722891566264


In [76]:
# логистическая регрессия

model = LogisticRegression()
model = model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))




Accuracy: 0.734375
Recall: 0.6162790697674418
Precision_score: 0.6022727272727273
f1: 0.6091954022988506


In [77]:
# дерево
tree = DecisionTreeClassifier(criterion='gini', max_depth=5, random_state=42)
cvs = cross_val_score(estimator=tree, X=X_st, y=y, cv=5, scoring='accuracy')
cvs1 = cross_val_score(estimator=tree, X=X_st, y=y, cv=5, scoring='recall')
cvs2 = cross_val_score(estimator=tree, X=X_st, y=y, cv=5, scoring='precision')
print('Accuracy:', cvs.mean())
print('Recall:', cvs1.mean())
print('Precision:', cvs2.mean())
print('F1_score:', (2 * cvs1.mean() * cvs2.mean())/(cvs1.mean() + cvs2.mean()))





Accuracy: 0.7461845344198285
Recall: 0.5825995807127883
Precision: 0.6551038843721771
F1_score: 0.616728092188623


In [78]:
# Норм. данные!

# Knn

X_train, X_test, y_train, y_test = train_test_split(X_norm, y, test_size=1/3, random_state=42)

knn = KNeighborsClassifier(7)

knn.fit(X_train,y_train)

y_pred = knn.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))



Accuracy: 0.7109375
Recall: 0.5465116279069767
Precision_score: 0.573170731707317
f1: 0.5595238095238095


In [79]:
# логист. регрессия

model = LogisticRegression()
model = model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print('Accuracy:', accuracy_score(y_test, y_pred))
print('Recall:', recall_score(y_test, y_pred))
print('Precision_score:', precision_score(y_test, y_pred))
print('f1:', f1_score(y_test, y_pred))





Accuracy: 0.7421875
Recall: 0.5348837209302325
Precision_score: 0.6388888888888888
f1: 0.5822784810126581


In [80]:
# дерево


tree = DecisionTreeClassifier(criterion='gini', max_depth=5, random_state=42)
cvs = cross_val_score(estimator=tree, X=X_norm, y=y, cv=5, scoring='accuracy')
cvs1 = cross_val_score(estimator=tree, X=X_norm, y=y, cv=5, scoring='recall')
cvs2 = cross_val_score(estimator=tree, X=X_norm, y=y, cv=5, scoring='precision')
print('Accuracy:', cvs.mean())
print('Recall:', cvs1.mean())
print('Precision:', cvs2.mean())
print('F1_score:', (2 * cvs1.mean() * cvs2.mean())/(cvs1.mean() + cvs2.mean()))




Accuracy: 0.7487819370172311
Recall: 0.5825995807127883
Precision: 0.6614478114478114
F1_score: 0.6195249795807436


In [82]:
# Кросс валидация

# KNN на норм данных тк они показали наил. результат

KNeighborsClassifier(7)
cvs = cross_val_score(estimator=knn, X=X_norm, y=y, cv=5, scoring='accuracy')
cvs1 = cross_val_score(estimator=knn, X=X_norm, y=y, cv=5, scoring='recall')
cvs2 = cross_val_score(estimator=knn, X=X_norm, y=y, cv=5, scoring='precision')
print('Accuracy:', cvs.mean())
print('Recall:', cvs1.mean())
print('Precision:', cvs2.mean())
print('F1_score:', (2 * cvs1.mean() * cvs2.mean())/(cvs1.mean() + cvs2.mean()))




Accuracy: 0.7552414905356082
Recall: 0.5487071977638016
Precision: 0.690031746031746
F1_score: 0.6113077943170474


In [83]:
# Логистич. регрессия для станд. данных тк они показали наил. результат

model = LogisticRegression()
cvs = cross_val_score(estimator=model, X=X_st, y=y, cv=5, scoring='accuracy')
cvs1 = cross_val_score(estimator=model, X=X_st, y=y, cv=5, scoring='recall')
cvs2 = cross_val_score(estimator=model, X=X_st, y=y, cv=5, scoring='precision')
print('Accuracy:', cvs.mean())
print('Recall:', cvs1.mean())
print('Precision:', cvs2.mean())
print('F1_score:', (2 * cvs1.mean() * cvs2.mean())/(cvs1.mean() + cvs2.mean()))

Accuracy: 0.7708853238265002
Recall: 0.5747030048916841
Precision: 0.715468528559464
F1_score: 0.6374065813848232


# Итоги:
В качестве оценки моделей используется F-мера

KNN дает наилучшие результаты при нормализованных и обычных данных: 0.5595238095238095 (без кросс валидации), 0.6113077943170474 (с кросс валидацией)

Логистическая регрессия дает наилучшый результат при стандартизированных данных: 0.6091954022988506 (без кросс валидации), 0.6374065813848232 (с кросс валидацией)

Дерево дает наилучный результат при обычных или нормализированных данных: 0.6195249795807436 (с кросс валидацией)

Наилучшей моделью я считаю Логистическую регрессию из-за наилучшего показатля F-меры